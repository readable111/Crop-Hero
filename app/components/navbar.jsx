import { useCallBack } from 'react';
import { StyleSheet, View, Text, Scrollable } from 'react-native'
import { useFonts } from 'expo-font'
import { Link } from 'expo-router';


const Navbar = () =>{ 
	const [fontsLoaded, fontError] = useFonts({
	'Domine-Regular': require('../../assets/fonts/Domine/Domine-Regular.ttf'),
	});

	if (!fontsLoaded && !fontError) {
		return null;
	}
		
	return(
	<View style = {styles.container}>
		<Text>This will be the Navbar Component</Text>
	</View>)
};

const styles = StyleSheet.create({
	container: {
		backgroundColor: 0x97A5BF,
		height: '100%',
	}
})

export default Navbar;
