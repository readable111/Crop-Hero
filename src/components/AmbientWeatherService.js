import axios from 'axios';

//Use the following URL to get a list of all devices:
//https://private-anon-f9255d0157-ambientweather.apiary-proxy.com/v1/devices?apiKey=1&applicationKey=2

//List of fields that MAY be returned depending on the device
//https://github.com/ambient-weather/api-docs/wiki/Device-Data-Specs

//List of used URLs
const BASE_URL = 'https://api.ambientweather.net/v1';
//Hex code for mock server may change over time as it is a free service provided by Apiary for Ambient Weather subscribers
//To Update: visit https://ambientweather.docs.apiary.io/, Switch to Console, select Query Device Data, provide necessary params, set to Mock Server, Call Resource, copy the base URL in the listed call
const BASE_MOCK_URL = 'https://private-anon-f9255d0157-ambientweather.apiary-mock.com/v1' 

//Timestamp of last request
//Must be global
let lastRequestTime = 0;
//Enforce rate limit of 1 request per second with slight gap to avoid AW's rate limiter
const rateLimiter = (func) => {
  return (...args) => {
    const now = Date.now();
    if (now - lastRequestTime < 1050) { // 1000 ms = 1 second
      console.warn("Rate limit exceeded, skipping request.");
      return Promise.resolve(null); // Ignore request and resolve with null
    }
    lastRequestTime = now;
    return func(...args);
  };
};

// Fetch data for a specific device or location with rate limiter function wrapping it to enforce limiting per Ambient Weather API rules
const fetchWeatherData = rateLimiter(async (apiKey, appKey, deviceMacAddress) => {
  try {
    //Use axios to make a REST request to the API for the data of a specific device
    const response = await axios.get(
      `${BASE_URL}/devices/${deviceMacAddress}`,
      {
        params: {
          apiKey: apiKey,
          applicationKey: appKey,
        },
        timeout: 1000
      }
    );

    //this if is only true if the GET succeeded (so app/api keys are valid & so is MAC) but no data was returned by the device
    if (!Array.isArray(response.data) || !response.data.length) {
      console.log("Mock called because device not reporting")
      const mockResponse = await axios.get(
        `${BASE_URL}/devices`,
        {
          params: {
            apiKey: apiKey,
            applicationKey: appKey,
          },
          timeout: 1000
        }
      );
      console.log("response retrieved")
      
      return mockResponse.data
    } else{ 
      return response.data;
    }
    
  } catch (error) {
    console.warn(error.data)
    if (!error.data) {
      throw "request failed";
    } else if (error._response) { //status code not in 2xx range
      throw error._response;
    } else if (error._request) { //no response received for request
      throw error._request;
    }else { //error when setting up request
      throw error._message;
    }
  }
});

function sleep(ms) {
  return new Promise(resolve => setTimeout(resolve, ms));
}

export { fetchWeatherData, sleep };