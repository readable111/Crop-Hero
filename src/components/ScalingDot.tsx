/****
 * @author Daniel Moreno
 * @reviewer Daniel Moreno
 * @tester Daniel Moreno
 ***/

import React from 'react';
import { StyleSheet, useWindowDimensions, ViewStyle, StyleProp } from 'react-native';
import Animated, {
  interpolate,
  Extrapolation,
  interpolateColor,
  useAnimatedStyle,
  SharedValue
} from 'react-native-reanimated';

type ScalingDotProps = {
  data: Array<Object>;
  scrollX: SharedValue<number>;
  dotStyle?: ViewStyle;
  inActiveDotOpacity?: number;
  inActiveDotColor?: string;
  activeDotScale?: number;
  activeDotColor?: string;
  style?: StyleProp<ViewStyle>;
};

type Props = Omit<ScalingDotProps, 'data' | 'containerStyle'> & {
  index: number;
};

const ScalingDot = ({
  scrollX,
  index,
  dotStyle,
  inActiveDotOpacity,
  inActiveDotColor,
  activeDotScale,
  activeDotColor,
}: Props) => {
  const dp = {
    inActiveDotColor: inActiveDotColor || '#347af0',
    activeDotColor: activeDotColor || '#347af0',
    animationType: 'scale',
    inActiveDotOpacity: inActiveDotOpacity || 0.5,
    activeDotScale: activeDotScale || 1.4,
  };
  const { width } = useWindowDimensions();
  const inputRange = [
    (index - 1) * width, 
    index * width, 
    (index + 1) * width
  ];
  const opacityOutputRange = [dp.inActiveDotOpacity, 1, dp.inActiveDotOpacity];
  const scaleOutputRange = [1, dp.activeDotScale, 1];
  const colorOutputRange = [
    dp.inActiveDotColor,
    dp.activeDotColor,
    dp.inActiveDotColor,
  ];
  const extrapolation = {
    extrapolateRight: Extrapolation.CLAMP,
    extrapolateLeft: Extrapolation.CLAMP,
  };

  const animatedStyle = useAnimatedStyle(() => ({
    opacity: interpolate(
      scrollX.value,
      inputRange,
      opacityOutputRange,
      extrapolation,
    ),
    backgroundColor: interpolateColor(
      scrollX.value,
      inputRange,
      colorOutputRange,
    ),
    transform: [
      {
        scale: interpolate(
          scrollX.value,
          inputRange,
          scaleOutputRange,
          extrapolation,
        ),
      },
    ],
  }));

  return (
    <Animated.View
      key={`dot-${index}`}
      style={[styles.dotStyle, dotStyle, animatedStyle]}
    />
  );
};

const styles = StyleSheet.create({
  dotStyle: {
    width: 7,
    height: 7,
    borderRadius: 5,
    marginHorizontal: 5,
  },
});

export default ScalingDot;