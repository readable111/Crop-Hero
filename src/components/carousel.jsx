/****
 * @author Tyler Bowen
 * @reviewer Daniel Moreno
 * @tester 
 ***/

import React from 'react'
import { View, FlatList, Dimensions, Text, StyleSheet, Image} from 'react-native'
import Colors from '../../assets/Color.js'
const { width } = Dimensions.get('window')

const Carousel = ({data}) => {
  return(
          <FlatList
            data={data}
            horizontal
            pagingEnabled
            showsHorizontalScrollIndicator={false}
            keyExtractor={(item, index) => index.toString()}
            renderItem={({ item }) => (
          <View style={styles.item}>
          <View syle = {styles.image}>
            <Image source = {require("../../assets/icons/Thermometer.png")} />
          </View>
          <View style = {styles.text}>
            <Text>Temperature: {item.temp}F</Text> 
            <Text>Feels Like: {item.perc}F</Text> 
          </View>
          </View>
        )}
      />
  )
}

const styles = StyleSheet.create({
  item: {
    width: width * .8,
    height: 90,
    backgroundColor: Colors.ALMOND_TAN,
    borderRadius: 10,
    marginTop: 20,
    marginHorizontal: 30,
    justifyContent: 'center',
    alignItems: 'center',
    flexDirection: 'row',
    paddingLeft: 40
  },
  text:{
    flex: 1,
    alignContent: 'center',
    justifyContent:'center',
    marginHorizontal: 30
  },
  image: {
    flex:1,
    height: 67,
    width: 31,
    margin: 20,
  }
});

const HomeCarousel = ({data}) =>{
    return(<Carousel data={data}/>)

}
export default HomeCarousel
