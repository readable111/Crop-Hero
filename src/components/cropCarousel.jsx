/****
 * @author Tyler Bowen, Daniel Moreno
 * @reviewer Daniel Moreno
 * @tester 
 * 
 * Secondary author (Daniel) added dark mode
 ***/

import React from 'react'
import { View, FlatList, Dimensions, Text, StyleSheet, Image, Pressable } from 'react-native'
import { useRouter } from 'expo-router'
import Colors from '../../assets/Color.js'
const { width } = Dimensions.get('window')

const Carousel = ({data,  isDarkMode=false}) => {
  const router  = useRouter();

  return(
          <FlatList
            data={data}
            horizontal
            pagingEnabled
            showsHorizontalScrollIndicator={false}
            keyExtractor={(item, index) => index.toString()}
            renderItem={({ item }) => (
        <Pressable onPress = {() => router.push({pathname:'crops/cropspage', params: item})}>
          <View style={[styles.item, isDarkMode && styles.itemDark]}>
            <Image source = {require("../../assets/icons/cropDefaultImage.png")}/>
            <Text>Name: {item[10]}</Text> 
            <Text>Medium: {item[18]}</Text>
            <Text>Location: {item[19]}</Text>
            <Text>Start Date: {new Date(item[13]).toISOString().slice(0,10)}</Text>
          </View>
        </Pressable>
        )}
    />
  )
}

const styles = StyleSheet.create({
  item: {
    width: 151,
    height: 232,
    backgroundColor: Colors.SCOTCH_MIST_TAN,
    borderRadius: 10,
    marginLeft: 10,
    marginRight: 10,
    justifyContent: 'center',
    alignItems: 'center',
    padding: 4,
  },
  itemDark: {
    backgroundColor: Colors.LICHEN,
  }
});

const CropCarousel = ({crops, isDarkMode=false}) =>{
    return(<Carousel data={crops} isDarkMode={isDarkMode}/>)

}
export default CropCarousel
