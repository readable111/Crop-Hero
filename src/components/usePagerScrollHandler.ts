/****
 * @author Daniel Moreno
 * @reviewer Daniel Moreno
 * @tester 
 ***/

import { ReanimatedEvent, useEvent, useHandler } from 'react-native-reanimated';

interface GeneralHandler<
  Event extends object,
  Context extends Record<string, unknown>,
> {
  (event: ReanimatedEvent<Event>, context: Context): void;
}

type GeneralHandlers<
  Event extends object,
  Context extends Record<string, unknown>,
> = Record<string, GeneralHandler<Event, Context> | undefined>;

function usePagerScrollHandler(
  handlers: GeneralHandlers<object, Record<string, unknown>>,
  dependencies?: any,
) {
  const { context, doDependenciesDiffer } = useHandler(handlers, dependencies);

  return useEvent(
    event => {
      'worklet';
      const { onPageScroll } = handlers;

      if (onPageScroll && event.eventName.endsWith('onPageScroll')) {
        onPageScroll(event, context);
      }
    },
    ['onPageScroll'],
    doDependenciesDiffer,
  );
}

export default usePagerScrollHandler;