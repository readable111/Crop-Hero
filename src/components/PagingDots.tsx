/****
 * @author Daniel Moreno
 * @reviewer Daniel Moreno
 * @tester Daniel Moreno
 ***/

import React from 'react';
import { SharedValue } from 'react-native-reanimated';
import { View, ViewStyle, StyleProp, StyleSheet } from 'react-native';
import ScalingDot from './ScalingDot';

type Props = {
  data: Array<Object>;
  scrollX: SharedValue<number>;
  dotStyle?: ViewStyle;
  inActiveDotOpacity?: number;
  inActiveDotColor?: string;
  activeDotScale?: number;
  activeDotColor?: string;
  style?: StyleProp<ViewStyle>;
};

const PagingDots = ({ data, style, ...rest }: Props) => {
  return (
    <View style={[styles.container, style]}>
      {data.map((_, index) => {
        return <ScalingDot key={`dot-${index}`} index={index} {...rest} />;
      })}
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    position: 'absolute',
    bottom: 20,
    flexDirection: 'row',
    alignSelf: 'center',
  },
});

export default PagingDots;