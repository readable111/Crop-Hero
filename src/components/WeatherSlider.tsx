/****
 * @author Daniel Moreno
 * @reviewer Daniel Moreno
 * @tester 
 ***/

import React from 'react';
import PagerView from 'react-native-pager-view';
import {
  Text,
  View,
  StyleSheet,
  SafeAreaView,
  useWindowDimensions,
  Dimensions,
  Image,
} from 'react-native';
import Animated, {
  interpolate,
  Extrapolation,
  useSharedValue,
  useDerivedValue,
} from 'react-native-reanimated';
import PagingDots from './PagingDots';
import usePagerScrollHandler from './usePagerScrollHandler';
import Colors from '../../assets/Color';
import icons from '../../assets/icons/Icons';
import { Col, Row } from '../../assets/Grid';
const { width } = Dimensions.get('window')

const AnimatedPagerView = Animated.createAnimatedComponent(PagerView);

const INTRO_DATA = [
  {
    key: '1',
    image: icons.thermometer_santa_gray,
    line1Label: 'Temperature',
    line1: '70°F',
    line2Label: 'Feels Like',
    line2: '80°F',
  },
  {
    key: '2',
    image: icons.rainfall_black,
    line1Label: 'Wind Speed',
    line1: '10mph',
    line2Label: 'Rainfall',
    line2: ' 50%',
  },
  {
    key: '3',
    image: icons.humidity_santa_gray,
    line1Label: 'Humidity',
    line1: '50%',
    line2Label: 'Soil Moisture',
    line2: ' 0.2wfv',
  }
];

export const WeatherSlider = ({isDarkMode=false, intro_data=INTRO_DATA}) => {
  const { width } = useWindowDimensions();
  const ref = React.useRef<PagerView>(null);
  const positionSharedValue = useSharedValue(0);
  const scrollOffsetSharedValue = useSharedValue(0);

  const scrollX = useDerivedValue(() => {
    const interpolatedValue = interpolate(
      positionSharedValue.value + scrollOffsetSharedValue.value,
      [0, intro_data.length],
      [0, intro_data.length * width],
      {
        extrapolateRight: Extrapolation.CLAMP,
      },
    );

    return interpolatedValue;
  });

  const handler = usePagerScrollHandler({
    onPageScroll: (e: any) => {
      'worklet';
      positionSharedValue.value = e.position;
      scrollOffsetSharedValue.value = e.offset;
    },
  });

  return (
    <SafeAreaView style={[styles.flex, isDarkMode && styles.flexDark]}>
      <AnimatedPagerView
        initialPage={0}
        ref={ref}
        style={styles.PagerView}
        onPageScroll={handler}
      >
        {intro_data.map(( item, key ) => (
          <View
            key={key}
          >
            <Row height={80} >
              <Col relativeColsCovered={3} alignItems='flex-end'>
              <Image 
                style={{width: 32,}}
                source={item.image}
              />
              </Col>
              <Col relativeColsCovered={6} alignItems='center'>
                <Text style={isDarkMode ? {color:Colors.WHITE_SMOKE} : {color:Colors.ALMOST_BLACK}}>{item.line1Label}: {item.line1}</Text>
                <Text style={isDarkMode ? {color:Colors.WHITE_SMOKE} : {color:Colors.ALMOST_BLACK}}>{item.line2Label}: {item.line2}</Text>
              </Col>
              <Col relativeColsCovered={3} alignItems='flex-start'>
              </Col>
            </Row>
          </View>
        ))}
      </AnimatedPagerView>
      <View style={styles.dotsContainer}>
        <View style={styles.dotContainer}>
          <PagingDots
            data={intro_data}
            scrollX={scrollX}
            style={{
              top: 30,
            }}
            activeDotColor = {Colors.MALACHITE}
            inActiveDotColor = {Colors.IRISH_GREEN}
            inActiveDotOpacity={0.8}
          />
        </View>
      </View>
    </SafeAreaView>
  );
}

const styles = StyleSheet.create({
  flex: {
    flex: 1,
    width: width * 0.8,
    backgroundColor: Colors.ALMOND_TAN,
    borderRadius: 12,
  },
  flexDark: {
    backgroundColor: Colors.IRIDIUM,
  },
  PagerView: {
    flex: 1,
  },
  center: {
    justifyContent: 'space-between',
    alignItems: 'center',
    flexDirection:'row',
    flex: 1,
    padding: 20,
    paddingTop: 0,
    paddingBottom: 0,
  },
  text: {
    fontSize: 30,
    color: 'black'
  },
  dotsContainer: {
    flex: 1,
    justifyContent: 'space-evenly',
    marginTop: -75,
  },
  dotContainer: {
    justifyContent: 'center',
    alignSelf: 'center',
  },
});
