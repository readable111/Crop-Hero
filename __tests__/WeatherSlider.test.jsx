import React from "react";
import {
    render,
    waitFor,
    screen,
    fireEvent,
} from "@testing-library/react-native";
import {WeatherSlider} from "../src/components/WeatherSlider"
import Colors from "../assets/Color";
import AsyncStorage from "@react-native-async-storage/async-storage";
import renderer from "react-test-renderer";

describe('<WeatherSlider/>', () =>{
    test('renders correctly', () =>{
        const tree = render(<WeatherSlider/>).toJSON();

        expect(tree).toMatchSnapshot();
    })
})