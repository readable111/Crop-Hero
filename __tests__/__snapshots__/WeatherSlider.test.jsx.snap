// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<WeatherSlider/> renders correctly 1`] = `
<h1>
  Letâ€™s work together
</h1>
`;
